### DISCUSSION TITLE
"Late" check for module support

### DISCUSSION THREAD

**Discussion by jbemmel** [ipspace/netlab#1486] [General]
https://github.com/ipspace/netlab/blob/7fb94e86cac09c6e359cd3196bf734714d517c19/netsim/modules/__init__.py#L116

The comment acknowledges that this is late in the process. I encountered this when adding ```stp``` support for OS10; the ```node_pre_transform``` hook gets called and reports an error, while the "real" error is that the STP module isn't (yet) supported by the device type to begin with

Can we do this check earlier? After expanding group data but before calling module hooks for the device would be better imho - I'd like to be able to assume module support has been validated for the device being operated on

**Comment by jbemmel** [comment]
![image](https://github.com/user-attachments/assets/b0b77e9f-1177-49da-afc9-431de89cbc0c)

Any reason this wouldn't work?

**Reply by ipspace** [reply]
DHCP module adding "dhcp" to the list of node modules if you ask for a DHCP client on a link?

**Comment by ipspace** [comment]
> I encountered this when adding `stp` support for OS10; the `node_pre_transform` hook gets called and reports an error, while the "real" error is that the STP module isn't (yet) supported by the device type to begin with

Maybe you should ask yourself "why did the node_pre_transform report an error". Before the data validation routines are called you have to assume the data is potentially corrupt (I told you that when discussing the LAG module).

**Reply by jbemmel** [reply]
Then maybe we should move that check from node_pre_transform to node_post_transform (and remove the ```node_pre_transform``` hook because that's all it does)

**Reply by jbemmel** [reply]
And there may be more cases like that: https://github.com/ipspace/netlab/pull/1488
